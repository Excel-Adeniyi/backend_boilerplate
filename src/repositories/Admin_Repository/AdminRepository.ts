import { Pool, QueryResult, RowDataPacket } from "mysql2/promise";
import { AdminDetails } from "../../entities/Admin_Entity/admin_details";
import { IAdminRepository } from "../../interfaces/Admin_Interface/i_admin_repository";
import { DatabaseError } from "../../errors/db.errors";



export class AdminRepository implements IAdminRepository {
    protected pool: Pool;

    constructor(pool: Pool) {
        this.pool = pool;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    loginAdmin(username: string, password: string): Promise<AdminDetails | RowDataPacket[] | QueryResult> {
        throw new Error("Method not implemented.");
    }
    async createAdmin(payload: AdminDetails): Promise<AdminDetails | RowDataPacket[] | QueryResult> {
        const countquery1 = 'SELECT COUNT (*) as count FROM account WHERE username = ?'
        const countquery2 = 'SELECT COUNT (*) as count FROM admin_info WHERE email = ?'
        const countquery3 = 'SELECT COUNT (*) as count FROM admin_info WHERE phone = ?'

        const insertquery1 = 'INSERT INTO account (username, password) VALUES (?,?)';
        const params1 = [payload.username, payload.password];
        const selectquery2 = 'SELECT id FROM account WHERE username = ? AND password = ? '
        const insertquery3 = 'INSERT INTO admin_info (name, phone, email, account_id) VALUES(? ,? ,? ,?)'
        const connection = await this.pool.getConnection();

        try {
            connection.beginTransaction()
            const [SELECTCOUNT] = await connection.execute(countquery1, [payload.username])
            const selectcounts = SELECTCOUNT as RowDataPacket[]
            if (selectcounts[0].count > 0) {
                connection.rollback()
                throw new DatabaseError('user already exist')
            }
            await connection.execute(insertquery1, params1);

            const [selectResults] = await connection.execute(selectquery2, params1)
            const selectedResults = selectResults as RowDataPacket[]

            const [SELECTCOUNT2] = await connection.execute(countquery2, [payload.email])
            const selectcounts2 = SELECTCOUNT2 as RowDataPacket[]
            if (selectcounts2[0].count > 0) {
                connection.rollback()
                throw new DatabaseError('email already exist')
            }
            const [SELECTCOUNT3] = await connection.execute(countquery3, [payload.phone])
            const selectcounts3 = SELECTCOUNT3 as RowDataPacket[]
            if (selectcounts3[0].count > 0) {
                connection.rollback()
                throw new DatabaseError('phone already exist')
            }
            const [insertResult2] = await connection.execute(insertquery3, [payload.name, payload.phone, payload.email, selectedResults[0].id])
            await connection.commit()
            return insertResult2 as RowDataPacket[]
        } catch (error) {
            await connection.rollback()
            throw new DatabaseError(error as string)
        } finally {
            await connection.release()
        }
    }

    async updateAdmin(id: number, password: string): Promise<AdminDetails | QueryResult> {
        const query = 'UPDATE account SET password = ? WHERE id = ?'
        const params = [password, id]
        const connection = await this.pool.getConnection();
        try {
            const [updateResults] = await connection.execute(query, params)
            await connection.commit()
            return updateResults
        } catch (error) {
            await connection.rollback()
            throw new DatabaseError(error as string)
        } finally {
            await connection.release()
        }
    }

}
