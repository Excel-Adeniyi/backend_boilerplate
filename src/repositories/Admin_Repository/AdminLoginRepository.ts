import { Pool, QueryResult, RowDataPacket } from "mysql2/promise";
import { AdminDetails } from "../../entities/Admin_Entity/admin_details";
import { AdminRepository } from "./AdminRepository";
import { DatabaseError } from "../../errors/db.errors";


export class AdminLoginRepository extends AdminRepository{
    // private pool
    constructor(pool: Pool) {
        super(pool);
        // this.pool = pool;
    }
    async loginAdmin(username: string): Promise<AdminDetails | RowDataPacket[]  | QueryResult> {
        const query = 'SELECT password FROM account WHERE username = ?'
        const params = [username]

        const connection = await this.pool.getConnection();
        try {
            const [loginResults] = await connection.execute(query, params)
            await connection.commit()
            const rows = loginResults as RowDataPacket[]
            const data = rows[0]
            return data as RowDataPacket[]
        } catch (error) {
            await connection.rollback()
            throw new DatabaseError(error as string)
        } finally {
            await connection.release()
        }
    }

    
}