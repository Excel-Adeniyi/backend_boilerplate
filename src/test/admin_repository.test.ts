
import { describe, it, expect } from '@jest/globals';
import pool from '../config/dbConfig'; // Adjust the import according to your project structure
import { AdminRepository } from '../repositories/Admin_Repository/AdminRepository';
import { AdminLoginRepository } from '../repositories/Admin_Repository/AdminLoginRepository';

describe('Admin Repository', () => {
    process.env.NODE_ENV = "test"
    afterAll(async () => {
        await pool.end()
    })
    describe('createAdmin', () => {

        it('should create a user', async () => {
            const adminRepo = new AdminRepository(pool)
            const userData = { username: 'testuser', password: 'testpass', name: "tester", phone: "09826352611", email: "test@test.com" };
            const user = await adminRepo.createAdmin(userData);

            expect(user).toMatchSnapshot(
                "DatabaseError:user already exist"
            )
        });
    });

    describe('login the admin account', () => {
        it('should find a password by username', async () => {
            const loginRepo = new AdminLoginRepository(pool)


            const user = await loginRepo.loginAdmin('testuser');
            console.log(user)
            expect(user).toMatchSnapshot(
                `[ { password: 'testerpasser' } ]`
            );
        });
    });

    describe("Update Admin Credentials", () => {
        it("should update the admin credentials by the password given to it", async () => {
            const adminRepo = new AdminRepository(pool)
            const updateAdmin = await adminRepo.updateAdmin(18, 'testerpasser')

            expect(updateAdmin).toMatchObject(
                {
                    affectedRows: 1
                }
            )
        })
    })
});
