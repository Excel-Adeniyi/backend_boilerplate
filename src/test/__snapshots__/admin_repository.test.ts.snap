// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin Repository Update Admin Credentials should update the admin credentials by the password given to it 1`] = `
{
  "affectedRows": 1,
  "changedRows": 1,
  "fieldCount": 0,
  "info": "Rows matched: 1  Changed: 1  Warnings: 0",
  "insertId": 0,
  "serverStatus": 2,
  "warningStatus": 0,
}
`;

exports[`Admin Repository createAdmin should create a user 1`] = `
{
  "affectedRows": 1,
  "changedRows": 0,
  "fieldCount": 0,
  "info": "",
  "insertId": 3,
  "serverStatus": 2,
  "warningStatus": 0,
}
`;

exports[`Admin Repository login the admin account should find a password by username 1`] = `
[
  {
    "password": "testpass",
  },
]
`;

exports[`Admin Repository login the admin account should find a user by username and password 1`] = `
[
  {
    "password": "testpass",
  },
]
`;
